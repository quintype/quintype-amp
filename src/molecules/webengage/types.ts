import { Config } from "../../types/config";

export interface WebEngageTypes {
  config: Config;
  buttonText?: string;
}
