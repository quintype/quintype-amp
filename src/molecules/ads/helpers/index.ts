import get from "lodash.get";

export const getAdPropsFromConfig = ({ config, adName }) => get(config, ["ampConfig", "doubleclick", adName], null);
export const isEmpty = (obj: object) => Object.keys(obj).length === 0;
export const getDataSlotFromUnitPath = (adProps) => {
  const unitPath = get(adProps, "unit-path", null);
  if (unitPath) {
    delete adProps["unit-path"];
    adProps["data-slot"] = unitPath;
  }
  return adProps;
};

// export const getAdPropsFromConfig = ({ config, adName }) => get(config, ["ampConfig", "ads", adName], null);
//
// config = {  // in the future, this is how ads will be in the config
//   ampConfig: {
//     ...
//     ads: {
//       "top-ad": {
//         type: "industrybrains",
//         width: 300,
//         height: 250,
//         "data-width": "300",
//         "data-height": "250",
//         "data-cid": "19626-3798936394"
//       },
//       "bottom-ad": {
//         width: "320",
//         height: "50",
//         type: "doubleclick",
//         "data-slot": "/35096353/amptesting/formats/sticky"
//       }
//     },
//     ...
//   }
// }
