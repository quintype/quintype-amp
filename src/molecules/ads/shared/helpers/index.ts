import get from "lodash.get";
import cloneDeep from "lodash.clonedeep";

const isEmpty = (obj: object) => Object.keys(obj).length === 0;

export const getPropsForDfpAd = ({ overridingProps, config, adName }) => {
  if (isEmpty(overridingProps)) {
    const clone = cloneDeep(config);
    const adPropsFromConfig = get(clone, ["ampConfig", "doubleclick", adName], null);
    const unitPath = get(adPropsFromConfig, "unit-path", null);

    if (!adPropsFromConfig) throw new Error(`No props defined for ${adName} in config`);
    if (!unitPath) throw new Error(`unit-path not defined for ad ${adName} in config`);
    adPropsFromConfig["data-slot"] = unitPath; // renames unit-path to data-slot
    delete adPropsFromConfig["unit-path"];
    return adPropsFromConfig;
  } else return overridingProps;
};
