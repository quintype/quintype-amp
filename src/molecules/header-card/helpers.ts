import get from "lodash.get";
import { GetHeaderComponentHelperTypes } from "./types";

export const getHeaderComponentConfig = ({ componentName, config }: GetHeaderComponentHelperTypes) => {
  const components = get(config, ["opts", "headerCardConfig", "components"], []);
  const match = components.find((component) => component.name.toLowerCase() === componentName);
  if (!match)
    throw new Error(`Error in "getHeaderComponentConfig" function. Could not find "${componentName}" in config`);
  return match.config;
};
