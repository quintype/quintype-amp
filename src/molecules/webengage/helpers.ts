import get from "lodash.get";

export const getWebengageConfig = ({ story, config }) => {
  const webengageConfig = get(config, ["ampConfig", "webengage"], null);
  if (!webengageConfig) return null;
  const { "license-code": licenseCode, "india-data-center": indiaDataCenter } = webengageConfig;
  const websiteUrl = config.publisherConfig["sketches-host"];
  const region = indiaDataCenter ? "in" : "us";
  const baseUrl = "${base}&eventName=Amp Article View&" + getAttributes({ story, config });
  const trackingCode = {
    vars: {
      licenseCode,
      region
    },
    requests: {
      "custom-attributes": {
        baseUrl
      }
    },
    triggers: {
      "custom-attributesTrigger": {
        on: "visible",
        request: "custom-attributes"
      }
    }
  };
  return {
    licenseCode,
    websiteUrl,
    trackingCode
  };
};

const getAttributes = ({ story, config }) => {
  const headline = story.headline || "";
  const storyContentId = story["story-content-id"] || "";
  const firstSection = get(story, ["sections", "0"], null);
  const category = firstSection ? getCatergoryFromSection(firstSection, config) : "";
  const subCategory = firstSection ? getSubCategoryFromSection(firstSection) : "";
  const author = story["author-name"] || "";
  const articleType = story.access || "";
  const tagsArr = get(story, "tags", []);
  const tagsStr = tagsArr.length ? tagsArr.map((tag) => tag.name).join(",") : "";
  const event = "pageview";

  return `ArticleTitle=${headline}&ArticleId=${storyContentId}&Category=${category}&SubCategory=${subCategory}&Author=${author}&ArticleType=${articleType}&tags=${tagsStr}&event=${event}`;
};

export const getCatergoryFromSection = (section, config) => {
  const category = "";
  const parentId = section["parent-id"];
  if (parentId) {
    // returns name of parent section or name of section
    const sections = config.publisherConfig.sections;
    const parentSection = sections.find((sec) => sec.id === parentId);
    return parentSection ? parentSection.name : section.name;
  }
  return category;
};
const getSubCategoryFromSection = (section) => (section["parent-id"] ? section.name : "");
