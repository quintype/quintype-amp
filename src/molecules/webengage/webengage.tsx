import React, { Fragment } from "react";
import get from "lodash.get";
import { withStoryAndConfig } from "../../context";
import { WebPush, WebPushWidget, WebengageSubscribeButton, Spacer, Analytics } from "../../atoms";
import { WebEngageTypes } from "./types";
import { getWebengageConfig } from "./helpers";

export const WebEngageBase = ({ story, config, buttonText, width, height, visibility }: WebEngageTypes) => {
  const webengageConfig = getWebengageConfig({ story, config });
  const isWebEngageEnabled = get(config, ["opts", "featureConfig", "webengage", "isEnabled"], null);

  if (!webengageConfig || !isWebEngageEnabled) return null;
  const { trackingCode, websiteUrl, licenseCode } = webengageConfig;

  return (
    <Fragment>
      <Spacer token="m" />
      <Analytics id="webengage" type="webengage" targets={trackingCode} />
      <WebPush
        id="amp-web-push"
        helper-iframe-url={`${websiteUrl}/api/amp-web-push-helper-frame.html?version=1`}
        permission-dialog-url={`${websiteUrl}/api/amp-permission-dialog-web-engage.html?version=1`}
        service-worker-url={`${websiteUrl}/api/amp-service-worker-web-engage.js?licensecode=${licenseCode}&version=1`}
      />
      <WebPushWidget visibility={visibility || "unsubscribed"} width={width || "350px"} height={height || "60px"}>
        <WebengageSubscribeButton on="tap:amp-web-push.subscribe" text={buttonText} />
      </WebPushWidget>
    </Fragment>
  );
};

/**
 * WebEngage Component uses the WebPush, WebPushWidget and WebengageSubscribeButton atomic components internally.
 *
 * ```js
 * <WebEngage />
 * ```
 *
 * @param {Object} props Object containing parameters passed to the render prop
 * @param {String} props.buttonText Optional. A String which shows on the button.
 * @param {String} props.width Optional. A String which specifies the width of the button.
 * @param {String} props.height Optional. A String which specifies the height of the button.
 *
 *  @category Molecules
 * @component
 */

export const WebEngage = withStoryAndConfig(WebEngageBase);
