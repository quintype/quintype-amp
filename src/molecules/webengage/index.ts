import { WebEngage } from "./webengage";
export { WebEngage };
