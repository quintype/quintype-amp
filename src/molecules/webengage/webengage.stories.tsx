import * as React from "react";
import { storiesOf } from "@storybook/react";
import { WebEngage } from "./webengage";
import { Layout } from "../../atoms";
import { textStory, config } from "../../__fixtures__";

storiesOf("Webengage", module)
  .addDecorator((story) => (
    <Layout story={textStory} config={config}>
      {story()}
    </Layout>
  ))
  .add("Default", () => <WebEngage />)
  .add("With custom text", () => <WebEngage buttonText="Subscribe!" />)
  .add("With custom attributes", () => <WebEngage buttonText="Subscribe!" width="500px" height="100px" />);
