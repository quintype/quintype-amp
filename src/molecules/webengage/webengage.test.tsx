import React from "react";
import { WebEngageBase } from "./webengage";
import { shallow } from "enzyme";
import { config } from "../../__fixtures__";
import cloneDeep from "lodash.clonedeep";
import { WebPush, WebPushWidget, WebengageSubscribeButton, Analytics } from "../../atoms";
import { story as vikatanStory } from "../../__fixtures__/vikatan/textStory";
import { storyWithManyJsEmbeds } from "../../__fixtures__/";
import { getWebengageConfig, getCatergoryFromSection } from "./helpers";

const vikatanConfig = cloneDeep(config);
vikatanConfig.publisherConfig["sketches-host"] = "https://www.vikatan.com";
vikatanConfig.publisherConfig.sections = [
  {
    "domain-slug": "sports",
    slug: "sports",
    name: "sports",
    "section-url": "https://sports.vikatan.com/",
    id: 8967,
    "parent-id": null,
    "display-name": "sports",
    collection: {
      slug: "sports",
      name: "விளையாட்டு ",
      id: 16494
    },
    data: null
  },
  {
    "domain-slug": null,
    slug: "technology",
    name: "technology",
    "section-url": "https://www.vikatan.com/technology",
    id: 8968,
    "parent-id": null,
    "display-name": "technology",
    collection: {
      slug: "technology",
      name: "டெக்னாலஜி ",
      id: 16495
    },
    data: null
  },
  {
    "domain-slug": null,
    slug: "anniversaries",
    name: "anniversaries",
    "section-url": "https://www.vikatan.com/anniversaries",
    id: 8969,
    "parent-id": null,
    "display-name": "anniversaries",
    collection: {
      slug: "anniversaries",
      name: "anniversaries",
      id: 16496
    },
    data: null
  }
];

const configWithoutWebEngage = cloneDeep(config);
delete configWithoutWebEngage.ampConfig.webengage;

const buggyConfig = cloneDeep(config);
buggyConfig.ampConfig.webengage["tracking-code"] = undefined;

describe("Webengage", () => {
  it("should match snapshot", () => {
    const wrapper = shallow(<WebEngageBase config={vikatanConfig} story={vikatanStory} />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render webengage component if valid config for it is provided", () => {
    const wrapper = shallow(<WebEngageBase config={vikatanConfig} story={vikatanStory} />);
    expect(wrapper.find(Analytics).exists()).toBeTruthy();
    expect(wrapper.find(WebPush).prop("helper-iframe-url")).toBe(
      "https://www.vikatan.com/amp-web-push-helper-iframe-url.html?version=1"
    );
    expect(wrapper.find(WebPush).prop("permission-dialog-url")).toBe(
      "https://www.vikatan.com/amp-web-push-permission-dialog-url.html?version=1"
    );
    expect(wrapper.find(WebPush).prop("service-worker-url")).toBe("https://www.vikatan.com/service-worker.js");
    expect(wrapper.find(WebPushWidget).children()).toHaveLength(1);
  });
  it("should not render webengage component if config not provided", () => {
    const wrapper = shallow(<WebEngageBase config={configWithoutWebEngage} story={vikatanStory} />);
    expect(wrapper.find(WebPush).exists()).toBeFalsy();
    expect(wrapper.find(WebPushWidget).exists()).toBeFalsy();
  });
  it("should pass custom text down to button", () => {
    const wrapper = shallow(<WebEngageBase config={vikatanConfig} buttonText="lorem ipsum" story={vikatanStory} />);
    expect(wrapper.find(WebengageSubscribeButton).prop("text")).toBe("lorem ipsum");
  });
});

describe("getWebengageConfig helper function", () => {
  it("should return null if webengage config not provided", () => {
    const webengageConfig = getWebengageConfig({ story: storyWithManyJsEmbeds, config: configWithoutWebEngage });
    expect(webengageConfig).toBeNull();
  });
  it("should return correct values", () => {
    const webengageConfig = getWebengageConfig({ story: vikatanStory, config: vikatanConfig });
    if (!webengageConfig) throw new Error("Unexpected condition in webengage test");
    const { trackingCode, host, licenseCode } = webengageConfig;
    const trackingCodeJson = JSON.stringify(trackingCode);
    const expectedTrackingCode =
      '{"vars":{"licenseCode":"~134105365","region":"us"},"requests":{"custom-attributes":{"baseUrl":"${base}&eventName=Amp Article View&ArticleTitle=ஆடைகளை ஊடுருவி படம் எடுக்கிறதா ஒன்ப்ளஸ் 8 ப்ரோ கேமரா... உண்மை என்ன? #VikatanAnalysis&ArticleId=ffa539a0-46f4-4309-8a13-dd1d6dfd8b6e&Category=technology&SubCategory=gadgets&Author=ம.காசி விஸ்வநாதன்&ArticleType=&tags=one plus,apple,technology,smart phones,gadgets,Spy Camera&event=pageview"}},"triggers":{"custom-attributesTrigger":{"on":"visible","request":"custom-attributes"}}}';
    expect(trackingCodeJson).toBe(expectedTrackingCode);
    expect(host).toBe("https://www.vikatan.com");
    expect(licenseCode).toBe("~134105365");
  });
});

describe("getCatergoryFromSection helper function", () => {
  it("should return correct category", () => {
    const mockConfig = {
      publisherConfig: {
        sections: [
          { name: "aaa", id: 111, "parent-id": null },
          { name: "bbb", id: 222, "parent-id": null },
          { name: "ccc", id: 333, "parent-id": null },
          { name: "ddd", id: 444, "parent-id": null },
          { name: "eee", id: 555, "parent-id": null }
        ]
      }
    };
    const sectionWithParentId1 = {
      "parent-id": 555,
      name: "Finance"
    };
    const sectionWithParentId2 = {
      "parent-id": 84302,
      name: "Finance"
    };
    const sectionWithoutParentId = {
      "parent-id": null,
      name: "abcd"
    };
    expect(getCatergoryFromSection(sectionWithParentId1, mockConfig)).toBe("eee");
    expect(getCatergoryFromSection(sectionWithParentId2, mockConfig)).toBe("Finance");
    expect(getCatergoryFromSection(sectionWithoutParentId, mockConfig)).toBe("");
  });
});
