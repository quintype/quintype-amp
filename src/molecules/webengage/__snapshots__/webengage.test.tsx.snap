// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webengage should match snapshot 1`] = `
<Fragment>
  <Spacer
    token="m"
  />
  <Analytics
    id="webengage"
    targets={
      Object {
        "requests": Object {
          "custom-attributes": Object {
            "baseUrl": "\${base}&eventName=Amp Article View&ArticleTitle=ஆடைகளை ஊடுருவி படம் எடுக்கிறதா ஒன்ப்ளஸ் 8 ப்ரோ கேமரா... உண்மை என்ன? #VikatanAnalysis&ArticleId=ffa539a0-46f4-4309-8a13-dd1d6dfd8b6e&Category=technology&SubCategory=gadgets&Author=ம.காசி விஸ்வநாதன்&ArticleType=&tags=one plus,apple,technology,smart phones,gadgets,Spy Camera&event=pageview",
          },
        },
        "triggers": Object {
          "custom-attributesTrigger": Object {
            "on": "visible",
            "request": "custom-attributes",
          },
        },
        "vars": Object {
          "licenseCode": "~134105365",
          "region": "us",
        },
      }
    }
    type="webengage"
  />
  <Component
    helper-iframe-url="https://www.vikatan.com/amp/api/v1/amp-web-push-helper-frame.html"
    id="amp-web-push"
    permission-dialog-url="https://www.vikatan.com/amp/api/v1/amp-permission-dialog-web-engage.html"
    service-worker-url="https://www.vikatan.com/amp/api/v1/amp-service-worker-web-engage.js"
  />
  <Component
    height="60px"
    visibility="unsubscribed"
    width="350px"
  >
    <WithTheme(Component)
      on="tap:amp-web-push.subscribe"
    />
  </Component>
</Fragment>
`;
